File: CSharpCodeBase/components/character/combat.lua
local Class = require 'utils.hump.class'
local Component = require 'components.component'

local CombatComponent = Class({__includes=Component, init=function(self, entity)
  Component.init(self, entity)
  self.comboTimeout = 0
  self.currentAction = 1
  self.entity.combat = self
end})

function CombatComponent:Attack(target, weapon)
  local actionTypes = self.entity.weaponContainer:GetActionTypes(self.target, weapon)
  
  if os.clock() > self.comboTimeout or self.currentAction > #actionTypes then
    self.currentAction = 1
  end
  
  self.entity.weaponContainer:SetComboState(self.currentAction, weapon)
  --print("Attack action type: " .. actionTypes[self.currentAction][1])
  self.entity.mecanim:SetFloat("action_type", actionTypes[self.currentAction][1])
  self.entity.mecanim:SetTrigger("action")    
  
  self.currentAction = self.currentAction + 1
  if self.currentAction > #actionTypes then
    self.currentAction = 1
  end
  self.comboTimeout = os.clock() + actionTypes[self.currentAction][2]
  
  self.entity.mover:LookAt(target:GetPosition())
end

function CombatComponent:ResetCombo()
  self.currentAction = 1
end

function CombatComponent:AttackVector(vec3, weapon)
  local actionTypes = self.entity.weaponContainer:GetActionTypes(nil, weapon)
  if os.clock() > self.comboTimeout or self.currentAction > #actionTypes then
    self.currentAction = 1
  end
  self.entity.weaponContainer:SetComboState(self.currentAction, weapon)
  self.entity.mecanim:SetFloat("action_type", actionTypes[self.currentAction][1])
  self.entity.mecanim:SetTrigger("action")    
  self.currentAction = self.currentAction + 1
  if self.currentAction > #actionTypes then
    self.currentAction = 1
  end
  self.comboTimeout = os.clock() + actionTypes[self.currentAction][2]
  self.entity.mover:LookAt(vec3)
end

return CombatComponent

File: CSharpCodeBase/components/character/combat.cs
using System;
 
 using System.Collections.Generic;
 
 using System.Linq;
 
 using System.Text;
 
 using UnityEngine;
 
 using System.Collections;
 
 using MainGame.core;
 
 using UnityEngine.EventSystems;
 
 namespace MainGame{
 public class combat :component{
 var CombatComponent = Class({__includes=Component, init=function(self, entity)
   Component.init(self, entity);
   this.comboTimeout = 0;
   this.currentAction = 1;
   this.entity.combat = self;
 }})
 public void CombatComponent:Attack(target, weapon){
   var actionTypes = this.entity.weaponContainer:GetActionTypes(this.target, weapon);
   
   if(os.clock() > this.comboTimeout  ||  this.currentAction > #actionTypes  ){ 
     this.currentAction = 1;
   }
   
   this.entity.weaponContainer:SetComboState(this.currentAction, weapon);
   //print("Attack action type: " .. actionTypes[this.currentAction][1])
   this.entity.mecanim:SetFloat("action_type", actionTypes[this.currentAction][1]);
   this.entity.mecanim:SetTrigger("action")    ;
   
   this.currentAction = this.currentAction + 1;
   if(this.currentAction > #actionTypes  ){ 
     this.currentAction = 1;
   }
   this.comboTimeout = os.clock() + actionTypes[this.currentAction][2];
   
   this.entity.mover:LookAt(target:GetPosition());
 }
 public void CombatComponent:ResetCombo(){
   this.currentAction = 1;
 }
 public void CombatComponent:AttackVector(vec3, weapon){
   var actionTypes = this.entity.weaponContainer:GetActionTypes(nil, weapon);
   if(os.clock() > this.comboTimeout  ||  this.currentAction > #actionTypes  ){ 
     this.currentAction = 1;
   }
   this.entity.weaponContainer:SetComboState(this.currentAction, weapon);
   this.entity.mecanim:SetFloat("action_type", actionTypes[this.currentAction][1]);
   this.entity.mecanim:SetTrigger("action")    ;
   this.currentAction = this.currentAction + 1;
   if(this.currentAction > #actionTypes  ){ 
     this.currentAction = 1;
   }
   this.comboTimeout = os.clock() + actionTypes[this.currentAction][2];
   this.entity.mover:LookAt(vec3);
 }
 Component}}